File: edu-ios/Sources/Models/RequestBodies.swift
import Foundation

// Encodable bodies for the write endpoints. Each carries a clientId so an
// offline replay is deduped server-side (IdempotencyKey).

func makeClientId() -> String { UUID().uuidString }

struct QuizAttemptBody: Encodable {
    let videoId: String?
    let courseId: String?
    let score: Int
    let total: Int
    let answers: [Int?]
    let clientId: String
}

struct VideoWatchedBody: Encodable {
    let videoId: String
    let clientId: String
}

struct ReviewGradeBody: Encodable {
    let questionId: String
    let correct: Bool
    let clientId: String
}

struct DrillSessionBody: Encodable {
    let slug: String
    let level: Int
    let total: Int
    let correct: Int
    let bestStreak: Int
    let mode: String
    let durationSec: Int
    var score: Int? = nil   // Rapid Fire points; omitted for practice
    let clientId: String
}

struct CredentialsBody: Encodable {
    let email: String
    let password: String
}

// Re-authentication for a destructive action (DELETE /me).
struct PasswordBody: Encodable {
    let password: String
}

struct RegisterBody: Encodable {
    let name: String
    let email: String
    let password: String
}

struct AppleName: Encodable {
    let givenName: String?
    let familyName: String?
}

struct AppleSignInBody: Encodable {
    let identityToken: String
    let fullName: AppleName?
}

File: edu-android/app/src/main/java/com/timpsonlyceum/lyceum/model/RequestBodies.kt
package com.timpsonlyceum.lyceum.model

import kotlinx.serialization.Serializable
import java.util.UUID

// Bodies for the write endpoints. Each carries a clientId so a replayed write is
// deduped server-side (IdempotencyKey) rather than counted twice.

fun makeClientId(): String = UUID.randomUUID().toString()

@Serializable
data class QuizAttemptBody(
    val videoId: String? = null,
    val courseId: String? = null,
    val score: Int,
    val total: Int,
    val answers: List<Int?>,
    val clientId: String,
)

@Serializable
data class VideoWatchedBody(val videoId: String, val clientId: String)

@Serializable
data class ReviewGradeBody(val questionId: String, val correct: Boolean, val clientId: String)

@Serializable
data class DrillSessionBody(
    val slug: String,
    val level: Int,
    val total: Int,
    val correct: Int,
    val bestStreak: Int,
    val mode: String,
    val durationSec: Int,
    /** Rapid Fire points; omitted for practice. */
    val score: Int? = null,
    val clientId: String,
)

@Serializable
data class CredentialsBody(val email: String, val password: String)

/** Re-authentication for a destructive action (DELETE /me). */
@Serializable
data class PasswordBody(val password: String)

@Serializable
data class RegisterBody(val name: String, val email: String, val password: String)

@Serializable
data class RefreshTokenBody(val refreshToken: String)

@Serializable
data class HandleBody(val handle: String)

@Serializable
data class DeletedResponse(val deleted: Boolean = false)

@Serializable
data class OkResponse(val ok: Boolean? = null)

/** A new lecture comment; `parentId` null means top-level. */
@Serializable
data class NewCommentBody(
    val videoId: String,
    val body: String,
    val parentId: String? = null,
)
